### ISSUE TITLE
elements overflowing past end of message

### ISSUE THREAD

**Issue by warmadmax** [indice-co/EDI.Net#208]
Can anyone see what i'm doing wrong with this class?
It appears to be overflowing and repeating some sections.
the results are in the attached file, 
[test.txt](https://github.com/indice-co/EDI.Net/files/7474303/test.txt)

i'm using version 1.9.18 via NuGet,
The file should end on the first END=5'
the overflow / repeating is happening at the following lines 

STL=2+Z+0000+1+1344+++++1344++1310+000+1344+1310'
which then repeats from ++++1344++1310+000+1344+1310'

i've tried the serializer with and without AutoEndSegmentsGroups enabled

```
{
    EdiSerializer ediserial = new EdiSerializer();
    ediserial.AutoEndSegmentGroups = true;
    ediserial.EnableCompression = true;
    ediserial.Serialize(txtWriter, grammar, edidata);
}

```

This is a Tradacoms9 invoice layout i'm specifying

```
public class TradacomsV9
{
       public class Interchange
    {
        public STX StartOfTransmission { get; set; } = new STX();

        public INVFIL INVFIL { get; set; } = new INVFIL();

        public List<INVOICE> INVOIC { get; set; } = new List<INVOICE>();

        public VATTLR VATTLR { get; set; } = new VATTLR();

        public INVTLR INVTLR { get; set; } = new INVTLR();

        [EdiValue("9(1)", Path = "END/0")]
        public int END_EdiMessagesCount { get; set; }
    }

    [EdiSegment]
    [EdiPath("STX")]
    public class STX
    {
        [EdiValue("X(4)", Path = "STX/0/0")]
        public string STDS_Identifier { get; set; } = "ANA";
        [EdiValue("9(1)", Path = "STX/0/1")]
        public string STDS_Version { get; set; } = "1"; // fixed

        [EdiValue("X(13)", Path = "STX/1/0")]
        public string FROM_SenderCode { get; set; } = "00001";

        [EdiValue("X(35)", Path = "STX/1/1")]
        public string FROM_SenderName { get; set; } = "A";

        [EdiValue("X(13)", Path = "STX/2/0")]
        public string UNTO_ReciptCode { get; set; } = "00002";

        [EdiValue("X(35)", Path = "STX/2/1")]
        public string UNTO_ReciptName { get; set; } = "B";

        [EdiValue("9(6)", Path = "STX/3/0", Format = "yyMMdd", Description = "TRDT - Date")]
        [EdiValue("9(6)", Path = "STX/3/1", Format = "HHmmss", Description = "TRDT - Time")]
        public DateTime TRDT_DateTime { get; set; }
        [EdiValue("X(14)", Path = "STX/4/0")]
        public string SNRF_TransmissionRef { get; set; } = "B2";

        // <EdiValue("X(8)", Path:="STX/5/0")>
        // Public Property RCRF_Password As String = ""

        [EdiValue("X(6)", Path = "STX/6/0")]
        public string APRF_AppRef { get; set; } = "INVFIL"; // INVFIL on example
    }

    // <EdiSegment, EdiPath("MHD")>
    // Public Class MHD
    // <EdiValue("9(12)", Path:="MHD/1/0")>
    // Public Property MSRF_MsgRef_Count As Integer = 1
    // <EdiValue("X(6)", Path:="MHD/2/0")>
    // Public Property TYPE_Type As String = "INVFIL" ' or INVOIC, or VATTLR
    // <EdiValue("9(1)", Path:="MHD/2/1")>
    // Public Property TYPE_Version As Integer = 9
    // End Class

    [EdiSegment]
    [EdiPath("MHD")]
    public class MHD
    {
        [EdiValue("9(1)", Path = "*/0/0")]
        public int Position { get; set; }
        public MHD_Type Type { get; set; } = new MHD_Type();
    }

    [EdiElement]
    [EdiPath("MHD/1")]
    public class MHD_Type
    {
        [EdiValue("X(6)", Path = "*/*/0")]
        public string MessageType { get; set; }
        [EdiValue("9(1)", Path = "*/*/1")]
        public int Version { get; set; }

        public override string ToString()
        {
            if (!string.IsNullOrWhiteSpace(MessageType))
                return $"{MessageType} v{Version}";
            return base.ToString();
        }
    }

    [EdiSegment]
    [EdiPath("MTR")]
    public class MTR
    {
        [EdiValue("9(1)", Path = "MTR/0")]
        public int NOSG_NumbOfSegmentsInMSG { get; set; } = 0;
    }

    [EdiMessage]
    [EdiCondition("INVFIL", Path = "MHD/1")]
    public class INVFIL
    {
        public MHD MHD { get; set; } = new MHD() { Type = new MHD_Type() { MessageType = "INVFIL", Version = 9 } };
        public TYP TYP { get; set; } = new TYP();
        public SDT SDT { get; set; } = new SDT();
        public CDT CDT { get; set; } = new CDT();

        public DNA DNA { get; set; } = new DNA();
        public FIL FIL { get; set; } = new FIL();

        // Public Property FDT As New FDT

        public MTR MTR { get; set; } = new MTR();
    }

    [EdiSegment]
    [EdiPath("TYP")]
    public class TYP
    {
        [EdiValue("9(4)", Path = "TYP/0")]
        public string TCDE_TransCode { get; set; } = "0700";
        [EdiValue("X(12)", Path = "TYP/1")]
        public string TTYP_TransType { get; set; } = "INVOICES";
    }

    [EdiSegment]
    [EdiPath("SDT")]
    public class SDT
    {
        [EdiValue("9(13)", Path = "SDT/0/0")]
        public string SIDN_SupplierEAN { get; set; } = "00001";
        [EdiValue("X(17)", Path = "SDT/0/1")]
        public string SIDN_SupplierIDbyCust { get; set; } = "1234";

        [EdiValue("X(40)", Path = "SDT/1")]
        public string SNAM_SupplierName { get; set; } = "A";
        [EdiValue("X(35)", Path = "SDT/2")]
        public string SADD_line1 { get; set; } = "A1";
        [EdiValue("X(35)", Path = "SDT/2/1")]
        public string SADD_Line2 { get; set; } = "A2";
        [EdiValue("X(35)", Path = "SDT/2/2")]
        public string SADD_Line3 { get; set; } = "A3";
        [EdiValue("X(35)", Path = "SDT/2/3")]
        public string SADD_Line4 { get; set; } = "A4";
        [EdiValue("X(35)", Path = "SDT/2/4")]
        public string SADD_PostCode { get; set; } = "A5";

        // <EdiValue("9(9)", Path:="SDT/4/0")>
        // Public Property VATN_VatRegNo As String = "123"
        [EdiValue("X(17)", Path = "SDT/3/1")]
        public string VATN_VatRegAlpha { get; set; } = "A123";
    }


    [EdiSegment]
    [EdiPath("CDT")]
    public class CDT
    {
        [EdiValue("9(13)", Path = "CDT/0/0")]
        public string CIDN_CustEAN { get; set; } = "00002";
        [EdiValue("X(17)", Path = "CDT/0/1")]
        public string CIDN_CustIDBySupplier { get; set; } = "B2";

        [EdiValue("X(40)", Path = "CDT/1")]
        public string CNAM_SupplierName { get; set; } = "B";
        [EdiValue("X(35)", Path = "CDT/2")]
        public string CADD_line1 { get; set; } = "A1";

        [EdiValue("X(17)", Path = "CDT/3/1")]
        public string VATR_VatRegAlpha { get; set; } = "B1";
    }

    [EdiSegment]
    [EdiPath("FIL")]
    public class FIL
    {
        [EdiValue("9(1)", Path = "FIL/0")]
        public string FLGN_FileGenNumnber { get; set; } = "-1";
        [EdiValue("9(1)", Path = "FIL/1")]
        public string FLVN_FileVersionNo { get; set; } = "-1";
        [EdiValue("9(6)", Path = "FIL/2", Format = "yyMMdd")]
        public DateTime FLDT_FileCreationDate { get; set; } = DateTime.Now;
    }

    [EdiMessage]
    [EdiCondition("INVOIC", Path = "MHD/1")]
    public class INVOICE
    {
        public MHD MHD { get; set; } = new MHD() { Type = new MHD_Type() { MessageType = "INVOIC", Version = 9 } };
        public CLO CLO { get; set; } = new CLO();
        public IRF IRF { get; set; } = new IRF();
        public PYT PYT { get; set; } = new PYT();

        public DNA DNA { get; set; } = new DNA();

        public ODD ODD { get; set; } = new ODD();

        public List<ILD> ILD { get; set; } = new List<ILD>();

        public List<STL> STL { get; set; } = new List<STL>();

        public TLR TLR { get; set; } = new TLR();
        public MTR MTR { get; set; } = new MTR();
    }


    [EdiSegment]
    [EdiPath("CLO")]
    public class CLO
    {
        [EdiValue("9(13)", Path = "CLO/0/0")]
        public string CLOC_CustANA { get; set; } = "0";
        [EdiValue("X(40)", Path = "CLO/1")]
        public string CNAM_CustName { get; set; } = "";
        [EdiValue("X(35)", Path = "CLO/2/0")]
        public string CADD_1 { get; set; } = "";
        [EdiValue("X(35)", Path = "CLO/2/1")]
        public string CADD_2 { get; set; } = "";
        [EdiValue("X(35)", Path = "CLO/2/2")]
        public string CADD_3 { get; set; } = "";
        [EdiValue("X(35)", Path = "CLO/2/3")]
        public string CADD_4 { get; set; } = "";
        [EdiValue("X(35)", Path = "CLO/2/4")]
        public string CADD_PC { get; set; } = "";
    }


    [EdiSegment]
    [EdiPath("IRF")]
    public class IRF
    {
        [EdiValue("X(16)", Path = "IRF/0")]
        public string INVN_InvoiceNo { get; set; } = "";
        [EdiValue("9(6)", Path = "IRF/1", Format = "yyMMdd")]
        public DateTime IVDT_DateOfInvoice { get; set; }
        [EdiValue("9(6)", Path = "IRF/2", Format = "yyMMdd")]
        public DateTime TXDT_TaxDate { get; set; }
    }

    [EdiSegment]
    [EdiPath("PYT")]
    public class PYT
    {
        [EdiValue("9(10)", Path = "PYT/0")]
        public string SEQA_SequenceNo { get; set; } = "";
        [EdiValue("X(40)", Path = "PYT/1")]
        public string PAYT_PaymentTerms { get; set; }
        [EdiValue("9(6)", Path = "PYT/2/0", Format = "yyMMdd")]
        public DateTime PAYD_PaymentDate { get; set; }
        [EdiValue("9(1)", Path = "PYT/2/1")]
        public int PAYD_SettlementDiscountPercent { get; set; } = 0;
    }

    [EdiSegment]
    [EdiPath("DNA")]
    public class DNA
    {
        [EdiValue("9(1)", Path = "DNA/0")]
        public int SEQA_SequenceNo { get; set; } = 1;

        [EdiValue("X(40)", Path = "DNA/3")]
        public string GNAR_GeneralNarrative1 { get; set; } = "GBP";
    }

    [EdiSegment]
    [EdiPath("ODD")]
    public class ODD
    {
        [EdiValue("9(1)", Path = "ODD/0")]
        public int SEQA_SequenceNo { get; set; } = 1;

        [EdiValue("X(17)", Path = "ODD/1/0")]
        public string ORNO_CustOrderNo { get; set; }
        [EdiValue("X(17)", Path = "ODD/1/1")]
        public string ORNO_SupplierOrderNo { get; set; }
        [EdiValue("9(6)", Path = "ODD/1/3", Format = "yyMMdd")]
        public DateTime ORNO_DateOrderReceived { get; set; }
    }


    [EdiSegment]
    [EdiSegmentGroup("ILD", SequenceEnd = "STL")]
    public class ILD
    {
        [EdiValue("9(1)", Path = "ILD/0")]
        public int SEQA_SequenceNo1 { get; set; } = 1;
        [EdiValue("9(1)", Path = "ILD/1")]
        public int SEQB_SequenceNo2 { get; set; } = 1;

        [EdiValue("9(13)", Path = "ILD/2/0")]
        public long SPRO_SuppliersEANCode { get; set; }
        [EdiValue("X(30)", Path = "ILD/2/1")]
        public string SPRO_SuppliersItemCode { get; set; }

        [EdiValue("9(13)", Path = "ILD/4/0")]
        public long CPRO_CustomersEANCode { get; set; }
        [EdiValue("X(30)", Path = "ILD/4/1")]
        public string CPRO_CustomersItemCode { get; set; }
        [EdiValue("9(1)", Path = "ILD/5")]
        public long UNOR_ConsumerUnitsInTradedUnit { get; set; } = 1;
        [EdiValue("9(1)", Path = "ILD/6")]
        public long QTYI_TradedUnitsInvoiced { get; set; } = -1;
        [EdiValue("9(1)V9(4)", Path = "ILD/7")]
        public decimal AUCT_CostPriceExVat { get; set; }
        [EdiValue("9(1)V9(4)", Path = "ILD/8")]
        public long LEXC_LineTotalExVat { get; set; } = -1; // this is AUCT(price) x QTYI(qty)
        [EdiValue("X(1)", Path = "ILD/9")]
        public string VATC_VatRateCode { get; set; } = "S";
        [EdiValue("9(1)V9(3)", Path = "ILD/10")]
        public decimal VATP_VatRatePercent { get; set; } = 20;
        [EdiValue("X(30)", Path = "ILD/13")]
        public string TDES_TradedUnitDescLine1 { get; set; } = "";
    }

    [EdiSegment]
    [EdiSegmentGroup("STL", SequenceEnd = "TLR")]
    public class STL
    {
        [EdiValue("9(1)", Path = "STL/0")]
        public int SEQA_SequenceNo1 { get; set; } = 1;

        [EdiValue("X(1)", Path = "STL/1")]
        public string VATC_VatRateCode { get; set; } = "S";

        [EdiValue("9(1)V9(3)", Path = "STL/2")]
        public decimal VATP_VatRatePercent { get; set; } = 20;
        [EdiValue("9(1)", Path = "STL/3")]
        public int NRIL_NoOfLineItemsWithThisCode { get; set; }
        [EdiValue("9(1)V9(2)", Path = "STL/4")]
        public decimal LVLA_LineSubtotalExVat { get; set; }

        [EdiValue("9(1)V9(2)", Path = "STL/9")]
        public decimal EVLA_ExtendedSubtotalExVat { get; set; }
        [EdiValue("9(1)V9(2)", Path = "STL/11")]
        public decimal ASDA_ExtendedSubtotalIncSettlementDisc { get; set; }
        [EdiValue("9(1)V9(2)", Path = "STL/12")]
        public decimal VATA_VatAmountPayable { get; set; }
        [EdiValue("9(1)V9(2)", Path = "STL/13")]
        public decimal APSE_PayableSubtotalBeforeSettlementDisc { get; set; }
        [EdiValue("9(1)V9(2)", Path = "STL/14")]
        public decimal APSI_PayableSubTotalWithSettlementDisc { get; set; }
    }

    [EdiSegment]
    [EdiPath("TLR")]
    public class TLR
    {
        [EdiValue("9(1)", Path = "TLR/0")]
        public int NSTL_NumberOfTLR { get; set; }


        [EdiValue("9(1)V9(2)", Path = "TLR/1")]
        public decimal LVLT_TotalExVat { get; set; }

        [EdiValue("9(1)V9(2)", Path = "TLR/6")]
        public decimal EVLT_TotalExVatExtended { get; set; }

        [EdiValue("9(1)V9(2)", Path = "TLR/8")]
        public decimal ASDT_TotalExVATIncSettlementDiscount { get; set; }
        [EdiValue("9(1)V9(2)", Path = "TLR/9")]
        public decimal TVAT_TotalVATPayable { get; set; }
        [EdiValue("9(1)V9(2)", Path = "TLR/10")]
        public decimal TPSE_TotalPayableBeforeSettlementDisc { get; set; }
        [EdiValue("9(1)V9(2)", Path = "TLR/11")]
        public decimal TPSI_TotalPayableIncSettlementDisc { get; set; }
    }


    [EdiMessage]
    [EdiCondition("VATTLR", Path = "MHD/1")]
    public class VATTLR
    {
        public MHD MHD { get; set; } = new MHD() { Type = new MHD_Type() { MessageType = "VATTLR", Version = 9 } };
        public List<VRS> VRS { get; set; } = new List<VRS>();
        public MTR MTR { get; set; } = new MTR();
    }

    [EdiSegment]
    [EdiPath("VRS")]
    public class VRS
    {
        [EdiValue("9(1)", Path = "VRS/0")]
        public int SEQA_SequenceNo1 { get; set; } = 1;

        [EdiValue("X(1)", Path = "VRS/1")]
        public string VATC_VatRateCode { get; set; } = "S";

        [EdiValue("9(1)V9(3)", Path = "VRS/2")]
        public decimal VATP_VatRatePercent { get; set; } = 20;
        [EdiValue("9(1)V9(2)", Path = "VRS/3")]
        public decimal VSDE_FileSubTotalAmountExVat { get; set; }
        [EdiValue("9(1)V9(2)", Path = "VRS/4")]
        public decimal VSDI_FileSubTotalAmountExVatIncSettlementDisc { get; set; }

        [EdiValue("9(1)V9(2)", Path = "VRS/5")]
        public decimal VVAT_FileVatSubTotal { get; set; }
        [EdiValue("9(1)V9(2)", Path = "VRS/6")]
        public decimal VPSE_FileSubTotalPayableExSettlementDisc { get; set; }
        [EdiValue("9(1)V9(2)", Path = "VRS/7")]
        public decimal VPSI_FileSubTotalPayableIncSettlementDisc { get; set; }
    }

    [EdiMessage]
    [EdiCondition("INVTLR", Path = "MHD/1")]
    public class INVTLR
    {
        public MHD MHD { get; set; } = new MHD() { Type = new MHD_Type() { MessageType = "INVTLR", Version = 9 } };
        public TOT TOT { get; set; } = new TOT();

        public MTR MTR { get; set; } = new MTR();
    }

    [EdiSegment]
    [EdiSegmentGroup("TOT")]
    public class TOT
    {
        [EdiValue("9(1)V9(2)", Path = "TOT/0")]
        public decimal FASE_FileTotalExVatExDisc { get; set; }

        [EdiValue("9(1)V9(2)", Path = "TOT/1")]
        public decimal FASI_FileTotalExVatIncDisc { get; set; }

        [EdiValue("9(1)V9(2)", Path = "TOT/2")]
        public decimal FVAT_FileVATTotal { get; set; }

        [EdiValue("9(1)V9(2)", Path = "TOT/3")]
        public decimal FPSE_FileTotalPayableExSettleDisc { get; set; }
        [EdiValue("9(1)V9(2)", Path = "TOT/4")]
        public decimal FPSI_FileTotalPayableIncSettleDisc { get; set; }
        [EdiValue("9(1)V9(2)", Path = "TOT/5")]
        public decimal FTNI_FileInvoiceCount { get; set; }
    }
}

```

**Comment by warmadmax** [comment]
just an update,

changing 
```
[EdiValue("9(1)V9(3)", Path = "STL/2")]
        public decimal VATP_VatRatePercent { get; set; } = 20;
```

to 
```
[EdiValue("9(2)V9(3)", Path = "STL/2")]
        public decimal VATP_VatRatePercent { get; set; } = 20;
```

has stopped it overflowing for me,
it's possible there's a bug in there for zero values supplied to a decimal ?

**Comment by cleftheris** [comment]
Can you try without setting the `AutoEndSegmentGroups` to true.

**Comment by warmadmax** [comment]
it appears to be working correctly on both now which is odd, even changing back to "9(1)V9(3)"
sorry to have wasted your time.
please feel free to use the above class as an example of a tradacoms9 invoice if it's useful to you.

**Comment by cleftheris** [comment]
This is not your fault. This setting is from an old contribution (pull request) I had accepted before stabilizing the serializer when it helped escape/close the Segment groups that where not fully mapped to CLR classes. The serializer has since been evolved so I did not realize that the setting was breaking implementations. I will add a deprecation attribute so It won't be used in the future. Thanks for pointing it out.
